tests/teststun.nim

```diff
@@ -0,0 +1,87 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.used.}
+
+import options, strutils
+import bearssl
+import ../webrtc/udp_connection
+import ../webrtc/stun/stun_connection
+import ../webrtc/stun/stun_message
+import ../webrtc/stun/stun_attributes
+import ../webrtc/stun/stun_utils
+import ./asyncunit
+
+proc newRng(): ref HmacDrbgContext =
+  HmacDrbgContext.new()
+
+suite "Stun message encoding/decoding":
+  test "Get BindingRequest + encode & decode with a set username":
+    let
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(udpConn, TransportAddress(AnyAddress), true, newRng())
+      msg = conn.getBindingRequest(username = "Test")
+      encoded = msg.encode(msg.getAttribute(AttrUsername))
+      decoded = StunMessage.decode(encoded)
+
+    check:
+      msg.msgType == decoded.msgType
+      msg.transactionId == decoded.transactionId
+      msg.getAttribute(AttrUsername) == decoded.getAttribute(AttrUsername)
+      msg.getAttribute(AttrICEControlling).isSome()
+      msg.getAttribute(AttrICEControlled).isNone()
+      Priority.decode(msg.getAttribute(AttrPriority).get()).priority == 0x7effffff
```

Comment:
What's this constant? Is it in the decoded msg?

Reply:
https://github.com/status-im/nim-webrtc/pull/9/files#diff-5337d70c0d648e8db7070f22c850c47b5012039666d93ced99c62ff59cbc1a54R54-R60.
For our specific case (public server), Priority is set to the maximum. And this maximum is `0x7effffff`.
I added a comment.

Reply:
Shouldn't we be able to do something like `msg.getAttribute(AttrPriority).get() == decoded.getAttribute(AttrPriority).get()`

Reply:
It seems like we might want to have an equals proc where we can just call `msg == decoded`.

Reply:
`StunMessage.attributes` is a `seq[StunRawAttribute]`
`StunRawAttribute` is a wrapper with the attribute type, the length of the attribute and the Attribute itself encoded in a `seq[byte]`.

Reply:
> But I'm a bit confused. Are we mixing the test of encoding/decoding a msg with the encoding/decoding of a Priority?
> Does Priority need encode/decode independent procs?

`StunMessage.attributes` is a `seq[StunRawAttribute]`.
`StunRawAttribute` is a wrapper with the attribute type, the length of the attribute and the Attribute itself encoded in a `seq[byte]`.
If you want to access the content of the attribute itself, you need to "decode" the StunRawAttribute.
That's why I use `Priority.decode()`.
But we can do better I think. What do you think of replacing getAttribute with this:
```nim
proc getAttribute[T](msg: StunMessage): Option[T] = # with T is an Attribute
  for attr in msg.attributes:
    if attr.attributeType == T.getAttributeType():
      return T.decode(attr.value)
  return none(T)
```

Reply:
Shouldn't the decode proc decode the attributes?

Reply:
I think what is confusing me here is that the test is mixing testing if a msg is created correctly with testing if it is encoded/decoded correctly.

Reply:
In theory, yes. In practice, it's complicated.
The reason it's done like that is because I use binary-serialization to encode and decode the different attribute.
And it doesn't work with a sequence of Attribute which are very different from each other to encode and decode.
RawStunAttribute is here as a step between those two.

Reply:
I changed getAttribute behavior. It's clearer now.